XenAdmin/Wizards/RollingUpgradeWizard/PlanActions/RpuUploadAndApplySuppPackPlanAction.cs

```diff
@@ -0,0 +1,275 @@
+                        alreadyApplied = true;
+                    }
+                    else
+                        throw new Exception(string.Format("{0}: {1}. {2}", host, problems[0].Title,
+                            problems[0].Description));
+                }
+            }
+            catch (Exception ex)
+            {
+                log.Error(string.Format("Precheck failed on host {0}", host.Name()), ex);
+                throw;
+            }
+        }
+
+        private void ApplySuppPack(IXenConnection connection, Session session, string suppPack, Pool_update update)
+        {
+            try
+            {
+                AddProgressStep(string.Format(Messages.UPDATES_WIZARD_APPLYING_UPDATE, suppPack, host.Name()));
+
+                var task = Pool_update.async_apply(session, update.opaque_ref, host.opaque_ref);
+                PollTaskForResultAndDestroy(connection, ref session, task);
+            }
+            catch (Failure f)
+            {
+                if (f.ErrorDescription.Count > 1 && f.ErrorDescription[0] == Failure.UPDATE_ALREADY_APPLIED)
+                {
+                    log.InfoFormat("The update {0} is already applied on {1}. Ignoring this error.", suppPack,
+                        host.Name());
+                    ReplaceProgressStep(string.Format(Messages.UPDATES_WIZARD_SKIPPING_UPDATE, suppPack,
+                        host.Name()));
+                }
+                else
+                    throw;
+            }
+        }
+
+        private void AfterApplySuppPack(Pool_update update)
+        {
+            var afterApplyGuidanceLists = update.after_apply_guidance;
+            if (afterApplyGuidanceLists != null && afterApplyGuidanceLists.Count > 0)
+            {
+                foreach (update_after_apply_guidance afterApplyGuidance in afterApplyGuidanceLists)
+                {
+                    var planAction = GetAfterApplyGuidancePlanAction(host, afterApplyGuidance);
+                    DelayedPlanActions.Add(planAction);
+                }
+            }
+        }
+
+        private void RemoveSuppPackFromMaster(Session session, Host master, string suppPack, Pool_update update)
+        {
+            var isLastHostInPool = hosts.IndexOf(host) == hosts.Count - 1;
+            if (isLastHostInPool)
+            {
+                try
+                {
+                    AddProgressStep(string.Format(Messages.UPDATES_WIZARD_REMOVING_UPDATES_FROM_POOL, suppPack, master.Name()));
+
+                    Pool_update.pool_clean(session, update.opaque_ref);
+                    if (!update.AppliedOnHosts().Any())
+                        Pool_update.destroy(session, update.opaque_ref);
+
+                    uploadedSuppPacks.Remove(master);
+                }
+                catch (Exception ex)
+                {
+                    log.Error(string.Format("Remove update file from master failed on host {0}", master.Name()), ex);
+                    throw;

```

Comment:
The removal of the update should be done as a best-effort, as we do it in the RemoveUpdateFileFromMasterPlanAction.